ssl/ech/ech_helper.c

```diff
@@ -11,5 +11,141 @@
+#endif
+        || !PACKET_get_net_2(&pkt, &pi_tmp) /* ciphersuite */
+        || !PACKET_get_1(&pkt, &pi_tmp) /* compression */
+        || (*exts = PACKET_data(&pkt) - shstart) == 0
+        || !PACKET_as_length_prefixed_2(&pkt, &extpkt)
+        || PACKET_remaining(&pkt) != 0)
+        return 0;
+    extlens = PACKET_remaining(&extpkt);
+    if (extlens == 0) /* not an error, in theory */
+        return 1;
+    while (PACKET_remaining(&extpkt) > 0 && done < 1) {
+        if (!PACKET_get_net_2(&extpkt, &etype)
+            || !PACKET_get_length_prefixed_2(&extpkt, &oneext))
+            return 0;
+        if (etype == TLSEXT_TYPE_ech) {
+            if (PACKET_remaining(&oneext) != 8)
+                return 0;
+            *echoffset = PACKET_data(&oneext) - shstart - 4;
+            *echtype = etype;
+#ifdef OSSL_ECH_SUPERVERBOSE
+            echlen = PACKET_remaining(&oneext) + 4; /* type/length included */
+#endif
+            done++;
+        }
+    }
+#ifdef OSSL_ECH_SUPERVERBOSE
+    OSSL_TRACE_BEGIN(TLS) {
+        BIO_printf(trc_out, "orig SH/ECH type: %4x\n", *echtype);
+    } OSSL_TRACE_END(TLS);
+    ossl_ech_pbuf("orig SH", (unsigned char *)sh, sh_len);
+    ossl_ech_pbuf("orig SH session_id", (unsigned char *)sh + sessid_offset,
+                  sessid_len);
+    ossl_ech_pbuf("orig SH exts", (unsigned char *)sh + *exts, extlens);
+    ossl_ech_pbuf("orig SH/ECH ", (unsigned char *)sh + *echoffset, echlen);
+#endif
+    return 1;
+}
+
+/*
+ * make up HPKE "info" input as per spec
+ * encoding is the ECHconfig being used
+ * encodinglen is the length of ECHconfig being used
+ * info is a caller-allocated buffer for results
+ * info_len is the buffer size on input, used-length on output
+ * return 1 for success, zero otherwise
+ */
+int ossl_ech_make_enc_info(unsigned char *encoding, size_t encoding_length,
+                           unsigned char *info, size_t *info_len)
+{
+    WPACKET ipkt = { 0 };
+    BUF_MEM *ipkt_mem = NULL;
+
+    if (encoding == NULL || info == NULL || info_len == NULL)
+        return 0;
+    if (*info_len < (sizeof(OSSL_ECH_CONTEXT_STRING) + encoding_length))
+        return 0;
+    if ((ipkt_mem = BUF_MEM_new()) == NULL
+        || !WPACKET_init(&ipkt, ipkt_mem)
+        || !WPACKET_memcpy(&ipkt, OSSL_ECH_CONTEXT_STRING,
+                           sizeof(OSSL_ECH_CONTEXT_STRING) - 1)
+        /*
+         * the zero valued octet is required by the spec, section 7.1 so
+         * a tiny bit better to add it explicitly rather than depend on
+         * the context string being NUL terminated
+         */
+        || !WPACKET_put_bytes_u8(&ipkt, 0)
+        || !WPACKET_memcpy(&ipkt, encoding, encoding_length)) {
+        WPACKET_cleanup(&ipkt);
+        BUF_MEM_free(ipkt_mem);
+        return 0;
+    }
+    *info_len = sizeof(OSSL_ECH_CONTEXT_STRING) + encoding_length;
+    memcpy(info, WPACKET_get_curr(&ipkt) - *info_len, *info_len);
+    WPACKET_cleanup(&ipkt);
```

Comment:
add a goto label here to simplify a bit this method : 
````
     rv = 1;
cleanup: 
     WPACKET_cleanup(...);
`````

Reply:
that's moot after the changes above from @mattcaswell (that line's gone) (once more:-)